feat: support vmdk import for vm instance creation

mas-who opened the pull request:
## Note
- This is currently a draft spike for importing VMDK support in the UI. The aim for now is to show case how to import the VMDK file by using the operations websocket from lxd.

## Done
- Add support for VMDK import to create VM instances.

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - TBD

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://lxd-ui-863.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/12561)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,328 @@
+    toastNotify.failure("Instance creation failed.", new Error(msg));
+  };
+
+  const handleFinish = () => {
```
```suggestion
  const invalidateCache = () => {
```

edlerd commented on src/util/helpers.tsx:
```diff
@@ -290,3 +290,92 @@ export const getClientOS = (userAgent: string) => {
   return null;
 };
+
+export const getSupportedArchitectures = (
+  instanceType: "container" | "virtual-machine",
+): string[] => {
+  if (instanceType === "container") {
+    return [
+      "i686",
+      "x86_64",
+      "armv7l",
+      "aarch64",
+      "ppc",
+      "ppc64",
+      "ppc64le",
+      "s390x",
+      "mips",
+      "mips64",
+      "riscv32",
+      "riscv64",
+      "armv6l",
+      "armv8l",
+      "loongarch64",
+    ];
+  }
+
+  return ["x86_64", "aarch64", "ppc64le", "s390x"];
+};
```
As we only are interested in the vm options, I think we can keep the four options as a constant in the ConvertToInstanceBtn.

edlerd commented on src/util/helpers.tsx:
```diff
@@ -290,3 +290,92 @@ export const getClientOS = (userAgent: string) => {
+  return ["x86_64", "aarch64", "ppc64le", "s390x"];
+};
+
+export const sendFileByWebSocket = (
```
Better keep this in a dedicated file with name matching the function?

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,328 @@
+    });
+
+    if (current === total) {
+      handleCloseModal();
+      navigate(`/ui/project/${project?.name}/instances`);
```
I think we should emit an info notification here to inform the upload completed and the import on the server is running in the background.

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,321 @@
+
+        const ws = new WebSocket(wsUrl);
+
+        ws.onopen = () => {
```
lol, I was just commenting not to use use effect for this, but rely on ws.onopen and at that second you pushed this change.

mas-who replied:
haha yeah I was going through the code and realised use effect is not needed

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,321 @@
+      </Button>
+      {isOpen && (
+        <Portal>
+          <Modal
```
Maybe split the modal and the upload logic into a dedicated component?

mas-who replied:
Sounds good, I've moved all util functions into `util/convertToInstance.tsx`, moved the form component into `ConvertToInstanceForm.tsx` and kept the upload logic in `ConvertToInstanceBtn.tsx`.

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,321 @@
+          <Modal
+            close={handleCloseModal}
+            className="convert-image-modal"
+            title="Convert from image file"
```
I wouldn't announce this feature as "Convert from image file". Imagine what a user would expect this feature under. Probably more a "Upload virtual machine image" or alike. Also, I think the confirm button should say "Upload" or "Upload and create" instead of "Convert".

On upload progress, I would change this to "Uploading file to LXD server" or something more brief and precise.

mas-who replied:
Yeah that sounds good, instead of "Upload virtual machine image", maybe "Upload virtual machine" could be better since the former makes me think an image would be created instead of an instance. wdyt?

edlerd replied:
Yes, agreed to the shorter copy

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,321 @@
+              />
+              <label htmlFor="">Conversion options</label>
+              <div className="conversion-options">
+                <Input
+                  {...formik.getFieldProps("formatConversion")}
+                  type="checkbox"
+                  label="Format"
+                  disabled={!formik.values.imageFile}
+                  checked={formik.values.formatConversion}
+                />
+                <Input
+                  {...formik.getFieldProps("virtioConversion")}
+                  type="checkbox"
+                  label="Virtio"
+                  disabled={!formik.values.imageFile}
+                  checked={formik.values.virtioConversion}
+                />
```
I think we need some help text here, as these are not self-explanatory to a user.

mas-who replied:
I've added some help text for both both options, let me know what you think.

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,321 @@
+                onChange={handleFileChange}
+              />
+              <Input
+                {...formik.getFieldProps("name")}
```
I think we should populate the name from the sanitized file name, wdyt?

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,321 @@
+  return JSON.stringify(instanceArgs);
+};
+
+const ConvertToInstanceBtn: FC = () => {
```
We need to hide this depending on the api extension for conversion to be available.

mas-who replied:
good catch!

edlerd commented on src/pages/instances/actions/ConvertToInstanceBtn.tsx:
```diff
@@ -0,0 +1,321 @@
+  const handleSuccess = (instanceName: string) => {
+    const message = (
+      <>
+        Created instance <strong>{instanceName}</strong>.
```
I think we should link the instance name. Not sure if we need the configure link.

mas-who replied:
The thinking here is to keep the behaviour consistent with uploading an instance in `UploadInstanceBtn`. The lxd-migrate tool also allows the user to configure the instance during the migration setup steps, which I think would align with this behaviour?

edlerd replied:
I see, makes sense.

I think we should always link the instance name. I was conditioned to click on the name and got confused why in this occasion it is not clickable.

edlerd reviewed (commented):
QA for the happy path looks fine, some comments on the interaction and small ideas for the code.

mas-who commented:
@edlerd I've also added a function `isImageTypeRaw` to determine if a image is `raw` format, and based on that we would remove the `format` conversion option. The approach is based on heuristics so it's not foolproof, but I think this should be fine since if we fail to catch a raw image type then worst case it will go through conversion. Let me know what you think?

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,181 @@
+                Format{" "}
+                <Icon
+                  name="information"
+                  title="Check this option to convert the image into raw format"
```
```suggestion
                  title="Convert the image into raw format. This can be skipped if the image is already in raw format to speed up the import."
```
Not sure if the 2nd part is actually true. I'm assuming it is?

mas-who replied:
Yes that's correct, the backend will simply skip the conversion step for raw images

edlerd replied:
Assuming the above is true, then the label of the checkbox should read "Convert the image into raw format" and the tooltip contain the 2nd sentence.

mas-who replied:
Okay I'm going to vertically stack the checkboxes since the label is quite long :+1:

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,181 @@
+                Virtio{" "}
+                <Icon
+                  name="information"
+                  title="Check this option if the image requires Virtio drivers for conversion"
```
```suggestion
                  title="Add Virtio drivers to the image. This is needed if..."
```
Again, I am not sure if this is actually true. Can we explain why this should be enabled or disabled?

edlerd replied:
Does this 2nd option depend on the first one to be checked? If so, we should disable it with a title tag stating the dependency in case the 1st option is disabled.

mas-who replied:
You can have both `format` and `virtio` conversion options. Agreed on the label :+1:

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+
+    if (e.currentTarget.files) {
+      const file = e.currentTarget.files[0];
+      const suffix = "-instance";
```
```suggestion
      const suffix = "-import";
```
Maybe better?

edlerd commented on src/util/instances.tsx:
```diff
@@ -63,3 +63,7 @@ export const truncateInstanceName = (
 };
+
+export const sanitizeInstanceName = (name: string): string => {
+  return name.replace(/[^A-Za-z0-9-]/g, "");
```
```suggestion
  return name.replace(/[^A-Za-z0-9-]/g, "-");
```
Removing spaces can produce hard to read names, maybe better use a dash?

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+    if (e.currentTarget.files) {
+      const file = e.currentTarget.files[0];
+      const suffix = "-instance";
+      const fileExtension = getFileExtension(file.name);
```
This function doesn't really return the file extension, but an array. Maybe better to bundle this logic together with  sanitize instance name. Instead of splitting the work over two functions.

mas-who replied:
`getFileExtension` returns a string? Not sure about lumping the logic with `sanitizeInstanceName` since not all instance names might need a split by file extension

edlerd replied:
What I meant is: from a function called `getFileExtension(input: string)` I'd expect to get a string back, not an array.

mas-who replied:
but `getFileExtension` returns a string, see below code for the function.

```
export const getFileExtension = (filename: string): string => {
  if (!filename.includes(".")) {
    return "";
  }

  return `.${filename.split(".").pop()}` || "";
};
```

mas-who replied:
Aah I see, no problem :+1:

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+      close={handleCloseModal}
+      className="convert-to-instance-modal"
+      title={
+        uploadState ? "Uploading image to LXD" : "Upload virtual machine image"
```
```suggestion
        uploadState ? "Uploading to LXD" : "Upload virtual machine"
```
As discussed, maybe better copy here without the "image" part.

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+          {...formik.getFieldProps("architecture")}
+          id="architecture"
+          label="Image architecture"
+          options={supportedVMArchOptions}
```
Can this be an architecture that the host does not support? We have some logic in the base image selector to figure out which architectures are supported by the LXD server. We might want to reuse it here.

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+              Add Virtio drivers to the image{" "}
+              <Icon
+                name="information"
+                title="This is needed if the image does not have Virtio drivers installed."
```
```suggestion
                title="Mandatory, if the image does not have Virtio drivers installed."
```

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+              Convert the image into raw format{" "}
+              <Icon
+                name="information"
+                title="This can be skipped if the image is already in raw format to speed up the import."
```
```suggestion
                title="Can be skipped if the image is already in raw format to speed up the import."
```

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+          type="checkbox"
+          label={
+            <>
+              Convert the image into raw format{" "}
```
```suggestion
              Convert to raw format{" "}
```

edlerd commented on src/pages/instances/forms/ConvertToInstanceForm.tsx:
```diff
@@ -0,0 +1,179 @@
+          type="checkbox"
+          label={
+            <>
+              Add Virtio drivers to the image{" "}
```
```suggestion
              Add Virtio drivers{" "}
```

edlerd reviewed (commented):
This looks pretty neat already.

Some more suggestions on copy. Seeing it in action always sparks more ideas.
Also, some notes on the file name to instance name functionality.

edlerd commented:
We discussed last week with piper to 
1. Merge the "Upload instance" and "Convert from image buttons". Think of a good description, so users can find both features from that caption
2. Use radio buttons to switch between the two form contents. Place them on the top of the modal. The two different forms are very similar, so switching would only add/remove a couple of fields. Implementation wise, we can keep the two forms in dedicated files, but include them dynamically based on the active radio.
3. We already had good ideas for the labels of the two radio buttons. @piperdeck should have a link to the design file with sketches and the copy. Though those sketches did not have the radio buttons yet. Though, the labels should translate easily into the radio btn idea. Important is that we list the file types -- maybe in a muted font as part of the label.

edlerd commented on src/pages/instances/forms/UploadInstanceBackupFileForm.tsx:
```diff
@@ -158,95 +136,96 @@ const UploadInstanceBtn: FC = () => {
+            id="instance-file"
+            name="instanceFile"
+            type="file"
+            label="Instance backup file"
```
Should this match the radio description, including the example file extensions? Might be better.

mas-who replied:
Actually, what do you think about removing the label for this input? Since choosing the file type already relates to the file selection? For older LXD versions, we can show the label "LXD backup archive (.tar.gz)" since the file type selector will be hidden.

edlerd replied:
Yes, I think that works well.

edlerd commented on src/pages/instances/forms/UploadInstanceFileModal.tsx:
```diff
@@ -0,0 +1,89 @@
+        <>
+          <label htmlFor="file-type">Choose file type</label>
+          <div id="file-type" className="file-type">
+            <RadioInput
+              label="LXD backup archive (.tar.gz)"
+              inline
+              checked={fileType === "instance-backup"}
+              onChange={() => setFileType("instance-backup")}
+            />
+            <RadioInput
+              label="External format (.qcow2, .vmdk, etc...)"
+              inline
+              checked={fileType === "external-format"}
+              onChange={() => setFileType("external-format")}
+            />
```
The styling looks a little busy, maybe placing them below each other will give this more order. Secondly, I think a bit more bottom spacing would be good for the two radios. See the `snapshot-schedule` CSS class for an example where we set some custom styling to make the radio placement a bit better.

edlerd replied:
We can also use the `u-sv1` or up to `u-sv3` helper classes, that might be better than introducing more custom css.

mas-who replied:
thanks I keep forgetting that we have these utility classes!

mas-who commented:
@edlerd this PR is now ready for review. I've implemented the following as discussed:
1. Combine "upload instance" button with the new vm conversion button. In the upload modal, a file type selection is done via radio inputs. The content of the modal will change depending on the file type selected.
2. For external format file selection, if a raw image is selected, then the "format" conversion option will be removed and disabled.
3. For older LXD versions, file type selection will be hidden so users can only upload LXD backup archive files.
4. Added some logic to make the upload modal fixed height for larger screens (if there is enough viewport height). The aim here is to reduce the amount of layout shift in the modal when selecting different file formats. For screens with limited viewport height, the modal height will be dynamic and the form content will be scrollable. Let me know what you think.

edlerd reviewed (commented):
This is great, some comments on the layout below. Will give a more thorough review tomorrow.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+    });
+  };
+
+  const handleConfirmUpload = (values: UploadExternalFormatFileFormValues) => {
```
```suggestion
  const handleSubmit = (values: UploadExternalFormatFileFormValues) => {
```
There is no confirm here, is there?

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+        );
+      })
+      .catch((e) => {
+        close();
```
```suggestion
```
I think we should keep this open, so the user could resubmit and won't lose the form status, but we should probably reset the upload state.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+  };
+
+  const handleUploadError = (error: Error) => {
+    toastNotify.failure("Image upload failed.", error);
```
I think we have to reset the upload state here as well.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+          sendFileByWebSocket(
+            ws,
+            formik.values.imageFile,
+            FILE_CHUNK_SIZE,
```
Shall we just hard code this in the function? Probably don't need to make this variable.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+
+        ws.onopen = () => {
+          sendFileByWebSocket(
+            ws,
```
Shall we move the creation of the ws into the function and pass in the wsUrl?

mas-who replied:
very good point! :+1:

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+
+    if (e.currentTarget.files) {
+      const file = e.currentTarget.files[0];
+      const suffix = "-import";
+      const fileExtension = getFileExtension(file.name);
+      const sanitisedFileName = sanitizeInstanceName(
+        file.name.split(fileExtension)[0],
+      );
+      const instanceName = truncateInstanceName(sanitisedFileName, suffix);
```
When I select a file without a file extension, named "foo", I get "f-import", would expect "foo-import" for that case.

I think we don't need the file extension, maybe simplify this to be 1 function, that receives a filename and returns an instance name?

The other upload type uses "imported" suffix and uses a slightly different way to calculate the name. With a generic function, we could reuse it over there.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+
+      // If the image is already in raw format, remove the format conversion option
+      // this will optimise the conversion process since raw vm images do not need to be converted
+      const isRawImage = await isImageTypeRaw(file);
+      setIsRawImage(isRawImage);
+      if (isRawImage) {
+        await formik.setFieldValue("formatConversion", false);
+      }
```
I don't fully understand why we set this field and then disable the selector for it.

Should we maybe only set the field to false and remove the `isRawImage` state and its side effect to disable the field? Disabling it seems rather drastic to me and might be experienced as not wanted by users.

mas-who replied:
Yeah that makes sense, I guess the tooltip also indicates that the option can be skipped if it's a raw image, will remove the disabling logic.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+      imageFile: null,
+      formatConversion: true,
+      virtioConversion: false,
+      architecture: supportedVMArchOptions(
+        settings?.environment?.architectures || [],
+      )[0].value,
```
Can we use the `archOptions` variable, defined below? Probably can move that up a bit in the file and use it here.

edlerd commented on src/util/uploadExternalFormatFile.tsx:
```diff
@@ -0,0 +1,150 @@
+
+export const FILE_CHUNK_SIZE = 1024 * 1024;
+
+export const supportedVMArchOptions = (hostArchitectures: string[]) => {
```
```suggestion
export const externalUploadArchitectureOptions = (hostArchitectures: string[]) => {
```
Maybe clearer?

mas-who replied:
hmm not sure about this one, since these values are VM architectures supported by LXD in general and does not only relate to VM image uploads.

edlerd replied:
I see, then the existing name seems correct.

edlerd commented on src/util/uploadExternalFormatFile.tsx:
```diff
@@ -0,0 +1,150 @@
+      reader.readAsArrayBuffer(chunk);
+    };
+
+    reader.onload = (event) => {
+      if (!file) {
+        return;
+      }
+
+      const result = event.target?.result;
+      let data: ArrayBuffer | ArrayBufferLike;
+      if (result && typeof result === "string") {
+        // Convert string to ArrayBuffer
+        data = new TextEncoder().encode(result).buffer;
+      } else {
+        data = result as ArrayBuffer;
+      }
+
+      ws.send(data);
+
+      sentFileSize += data.byteLength;
+      if (onProgress) {
+        onProgress(sentFileSize, file.size);
+      }
+
+      readNextFileChunk();
+    };
+
+    reader.onerror = (e) => {
+      if (onError) {
+        onError(
+          e.target?.error
+            ? e.target.error
+            : new Error("Failed to read read file"),
+        );
+      }
+    };
+
+    readNextFileChunk();
+  }
+};
```
I found this hard to understand. Maybe splitting the responsibilities into a file reader, that just reports chunks and upload state, and another part that is hooking that up with the socket? Or maybe that is even worse, and the complexity is inherent.

mas-who replied:
I agree this is not the easiest code to understand but I am also not sure if we can split it like that since the socket and reader logic are tightly coupled. perhaps I can add some comments above the function describing how it works?

edlerd replied:
Yeah, a comment might help.

mas-who replied:
I've added some comments explaining the function.

edlerd reviewed (commented):
This works nicely and the design and layout now also looks very clean.

I think the error handling needs a bit of refinement. We just close the modal in both upload types, maybe we can keep it open not to lose the state? The user might want to retry, and preserving the state they created might be a better experience.

Apart from that a couple of small suggestions and ideas below.

mas-who commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,333 @@
+            id="architecture"
+            label="Image architecture"
+            options={archOptions}
+            disabled={!formik.values.imageFile || archOptions.length < 2}
```
also adding the message to all other disabled inputs

mas-who commented on src/sass/_upload_instance_modal.scss:
```diff
@@ -1,6 +1,19 @@
-  .p-modal__dialog {
-    margin: auto;
-    width: 35rem;
+  .conversion-options {
+    display: flex;
+    gap: 1rem;
+    margin-bottom: $spv--medium;
+  }
+
+  .fixed-height {
+    height: 33rem;
+  }
```
`fixed-height` is used in `UploadExternalFormatFileForm` and `UploadInstanceBackupFileForm`.

mas-who commented on src/pages/instances/forms/UploadInstanceFileModal.tsx:
```diff
@@ -0,0 +1,91 @@
+            </div>
+            <div className="u-sv3">
+              <RadioInput
+                label="External format (.qcow2, .vmdk, etc...)"
```
Yup, those are QCOW, QCOW2, VDI, VHDX, or VMDK detailed in the [docs](https://documentation.ubuntu.com/lxd/en/latest/howto/import_machines_to_instances/). Will add to the label, thanks for the `<abbr>` tip!

edlerd commented on src/util/uploadExternalFormatFile.tsx:
```diff
@@ -0,0 +1,156 @@
+    };
+
+    reader.onload = (event) => {
+      if (!file) {
+        return;
+      }
```
Can this guard be at the top of the function, so we have it only once?

mas-who replied:
unfortunately not, the external guard does not apply inside the context of the callbacks

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,334 @@
+            title={noFileSelectedMessage}
+          />
+          <label htmlFor="">Conversion options</label>
+          <Input
+            {...formik.getFieldProps("formatConversion")}
+            type="checkbox"
+            label={
+              <>
+                Convert to raw format{" "}
+                <Icon
+                  name="information"
+                  title={
+                    noFileSelectedMessage
+                      ? noFileSelectedMessage
+                      : "Can be skipped if the image is already in raw format to speed up the import."
+                  }
+                />
+              </>
+            }
+            disabled={!!noFileSelectedMessage}
+            checked={formik.values.formatConversion}
+          />
```
There is some clipping when the checkboxes are active. The blue border is cut on the left-hand side.

![Screenshot from 2024-09-11 18-06-07](https://github.com/user-attachments/assets/f138288c-0292-4452-9230-ea338f94ffac)
![Screenshot from 2024-09-11 18-06-02](https://github.com/user-attachments/assets/2a1b3f98-4f4f-488e-acc7-d50fb6f57c54)

mas-who replied:
had to give the checkboxes a small `margin-left`

edlerd replied:
This is a visual design crime, please don't misalign them with the rest of the modal. I think the root cause is the scrollable container in `UploadExternalFormatFileForm`. It also creates a half modal scrolling, see screen below. The same is true for the backup file. Maybe we should raise it to the `UploadInstanceFileModal` and not have scrollable container in the form components.

![image](https://github.com/user-attachments/assets/d36e7dcd-6040-4a90-9e45-0c86426ecdfb)

mas-who replied:
I see, the issue with that is the modal buttons are inside the form components since they need access to the form submission handlers. Raising the scrollable container to the modal level would defeat the point since the aim is to always show the modal buttons and make form content scrollable. I can create a component for the file type radio selection and move that into the form components to deal with the half scrolling form problem. I think some padding on the form component will probably fix the check boxes without misalignment.

edlerd replied:
Thanks for explaining, I now see the problem with that. Agree, your proposed solution sounds good.

mas-who replied:
I've made the above changes, let me know what you think? :slightly_smiling_face:

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,334 @@
+        sendFileByWebSocket(
+          wsUrl,
+          formik.values.imageFile,
+          1024 * 1024,
```
```suggestion
```
I think there was a misunderstanding. I meant keep the constant, but reference the constant directly in the function. In my opinion, we don't need it in the function parameters.

mas-who replied:
aah I see, done

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,334 @@
+      imageFile: null,
+      formatConversion: true,
+      virtioConversion: false,
+      architecture: archOptions[0].value,
```
I think If the settings are not loaded, the 0 index will cause an error, because we try to read `.value` from undefined.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,334 @@
+      // If the image is already in raw format, remove the format conversion option
+      // this will optimise the conversion process since raw vm images do not need to be converted
+      const isRawImage = await isImageTypeRaw(file);
+      if (isRawImage) {
+        await formik.setFieldValue("formatConversion", false);
+      }
```
```suggestion

      await formik.setFieldValue("formatConversion", !isRawImage);
```
I think we should also enable it in case the image is not raw. Even though that is usually enabled as it is the default.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,334 @@
+  }, [formik.resetForm, close, notify]);
+
+  const noFileSelectedMessage = !formik.values.imageFile
+    ? "A file must be selected first."
```
```suggestion
    ? "Please select a file before adding custom configuration."
```

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,334 @@
+              </>
+            }
+            disabled={!!noFileSelectedMessage}
+            checked={formik.values.formatConversion}
```
```suggestion
            checked={formik.values.formatConversion}
            title={noFileSelectedMessage}
```
We should have the title on the checkboxes as well.

mas-who replied:
for some reason when the input is a checkbox, `title` has no effect if specified on the `Input` component directly. Instead I added `title` attribute to the label.

edlerd replied:
I think this is because the checkbox gets wrapped by the label.

edlerd reviewed (commented):
Thanks for the adjustments!

If found one probably last visual flaw and a couple of very tiny suggestions to improve.

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,346 @@
+      // If the image is already in raw format, remove the format conversion option
+      // this will optimise the conversion process since raw vm images do not need to be converted
+      const isRawImage = await isImageTypeRaw(file);
+      if (isRawImage) {
+        await formik.setFieldValue("formatConversion", !isRawImage);
+      }
```
```suggestion
      await formik.setFieldValue("formatConversion", !isRawImage);
```
I think both cases are important, for non-raw images we should also enable conversion. This is the default, but a user might change between raw and non-raw files.

mas-who replied:
of course, that was my bad for not removing the condition check

edlerd commented on src/sass/_upload_instance_modal.scss:
```diff
@@ -1,6 +1,17 @@
+  }
+
+  .content-details {
+    padding-left: $sph--x-small;
```
```suggestion
    padding-left: $sph--x-small;
    margin-left: -$sph--x-small;
```
Better keep the content aligned with the title in the header

mas-who replied:
neat!

edlerd commented on src/pages/instances/forms/InstanceFileTypeSelector.tsx:
```diff
@@ -0,0 +1,45 @@
+}
+
+const InstanceFileTypeSelector: FC<Props> = ({ value, onChange, hide }) => {
+  if (hide) {
+    return null;
+  }
```
You can do it this way, personally prefer this decision in the parent to keep the interface clean. Like `{isConditional && <InstanceFileTypeSelector />}`

mas-who replied:
agreed, that's better :+1:

edlerd commented on src/pages/instances/forms/UploadInstanceFileModal.tsx:
```diff
@@ -0,0 +1,71 @@
+          setUploadState={setUploadState}
+          fixedFormHeight={fixedFormHeight}
+          fileType={fileType}
+          onFileTypeChange={setFileType}
```
```suggestion
          setFileType={setFileType}
```
Maybe better?

edlerd commented on src/pages/instances/forms/UploadInstanceFileModal.tsx:
```diff
@@ -0,0 +1,71 @@
+          setUploadState={setUploadState}
+          fixedFormHeight={fixedFormHeight}
+          fileType={fileType}
+          onFileTypeChange={setFileType}
```
```suggestion
          setFileType={setFileType}
```
Maybe better?

edlerd reviewed (approved):
QA looks good, design as well, few very tiny suggestions below.

edlerd reviewed (dismissed):
Good work, LGTM code and QA wise.

piperdeck commented:
This is a nitpick, but is there a reason we're using so much vertical space here? I think the modal might look a little bit better if its minimum vertical size was a bit smaller than this...

![CleanShot 2024-09-15 at 21 10 21@2x](https://github.com/user-attachments/assets/442b4f0b-7447-4701-b2b9-37d47b159491)

I'm noticing that we're not doing a lot to filter or validate the file extension of the file to be uploaded. Was this intentional? Seems like it might be a good idea to check and make sure the file that the user attaches at least has the correct extension, or maybe even set that as a parameter in the file upload dialog?

https://github.com/user-attachments/assets/36418b0d-18ca-4f19-aef6-2c8d719d9586

For example, when uploading pictures to Google Photos, the file picker is invoked in such a way as to disable anything that doesn't have an accepted file extension:

https://github.com/user-attachments/assets/aff5ac7a-b23b-40dc-8d75-9c861403fcfd

mas-who commented:
> This is a nitpick, but is there a reason we're using so much vertical space here? I think the modal might look a little bit better if its minimum vertical size was a bit smaller than this...
> 
> ![CleanShot 2024-09-15 at 21 10 21@2x](https://private-user-images.githubusercontent.com/45367207/367634597-442b4f0b-7447-4701-b2b9-37d47b159491.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0NzI0MjIsIm5iZiI6MTcyNjQ3MjEyMiwicGF0aCI6Ii80NTM2NzIwNy8zNjc2MzQ1OTctNDQyYjRmMGItNzQ0Ny00NzAxLWIyYjktMzdkNDdiMTU5NDkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDA3MzUyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkN2I3MjUwN2E1NDU5MTVhZDdlYzIyOTBkMTIxZWQzMGEyZjI5MDFhYjViNWI0MTE4YzIyZjBlZTgxNDllNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GYBlANAeq-5sSWe_JZJUaslL0_4BD8qldm__79OGCoE)
> 

This is done so that the modal does not change size depending on the file type selected. For the backup archive file type, there are less options to configure for the upload, so it appears to not utilise all the available space of the modal. Do you think we should rather have the modal adjust its size to the available content?

> I'm noticing that we're not doing a lot to filter or validate the file extension of the file to be uploaded. Was this intentional? Seems like it might be a good idea to check and make sure the file that the user attaches at least has the correct extension, or maybe even set that as a parameter in the file upload dialog?
> 
>  CleanShot.2024-09-15.at.21.21.33.mp4 
> For example, when uploading pictures to Google Photos, the file picker is invoked in such a way as to disable anything that doesn't have an accepted file extension:
> 
>  CleanShot.2024-09-15.at.21.25.04.mp4

For the file type extension, we can certainly restrict it during upload. However, the files does not necessarily need an extension for it to be valid. For example, a file named `lxd_vmdk_vm` or `lxd_vmdk_vm.vmdk` can both be a VMDK image. On the extreme side, a user can literally name their file `lxd_vmdk_vm.png` and it could still be a valid VMDK image. In this situation, we rely on the LXD backend to determine if a file uploaded is valid since it directly checks the binary content of the file. We would then show the error returned from the server if the file is invalid. Restricting file selection by extension may result in the UI blocking selection for valid files.

edlerd commented:
> For the file type extension, we can certainly restrict it during upload. However, the files does not necessarily need an extension for it to be valid. For example, a file named `lxd_vmdk_vm` or `lxd_vmdk_vm.vmdk` can both be a VMDK image. On the extreme side, a user can literally name their file `lxd_vmdk_vm.png` and it could still be a valid VMDK image. In this situation, we rely on the LXD backend to determine if a file uploaded is valid since it directly checks the binary content of the file. We would then show the error returned from the server if the file is invalid. Restricting file selection by extension may result in the UI blocking selection for valid files.

I think there is a way to have the file picker set to a list of allowed file extensions, but the user can still switch the filter to "any file type" if they really want. I think having the filter on the right file type might improve the experience.

edlerd commented:
I.e. choosing a file from a download directory with dozens or hundreds of files will be much easier with a filter on file extension applied.

mas-who commented:
> I.e. choosing a file from a download directory with dozens or hundreds of files will be much easier with a filter on file extension applied.

Yeah fair, I think then for instance backup files, we will filter file extensions to `.tar.bz2`, `.tar.gz`, `.tar.xz`, `.tar.lzma`, `.tar`, `.squashfs`, `.qcow2` and `.tar.zst` as per supported compression listed in LXD [source](https://github.com/canonical/lxd/blob/e6a6119184f3d98be0361abae7a5ff9f9389569b/shared/archive.go#L25). For external format file uploads, we will restrict to `.img`, `.qcow`, `.qcow2`, `.vdi`, `.vhdx` and `.vmdk` extensions as listed in the `lxd-migrate` [docs](https://documentation.ubuntu.com/lxd/en/latest/howto/import_machines_to_instances/).

piperdeck commented:
@mas-who I think I'm liking the fully responsive behaviour of the modal. What are your thoughts on it?

mas-who commented:
> @mas-who I think I'm liking the fully responsive behaviour of the modal. What are your thoughts on it?

@piperdeck I'm happy with this behaviour if you are happy! :)

mas-who commented:
@edlerd made the following changes based on feedback from @piperdeck :

1. adjusted the `accept` attribute for instance backup file picker. Turns out the support for this attribute on safari is not great, multi-segment file extensions such as `tar.gz`, `tar.bz2` have issues getting recognised resulting in the file picker to filter out those file extensions on safari. Instead, using MIME types should resolve this issue.

2. If an instance name is set on the create instance page, opening the file upload modal will automatically fill out the instance name input field. We do not generate an instance name based on the file name in this scenario

piperdeck commented:
Looks good to go!

edlerd commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,335 @@
+        />
+      </Form>
+      <footer className="p-modal__footer" id="modal-footer">
+        <Button
+          appearance="base"
+          className="u-no-margin--bottom"
+          type="button"
+          onClick={handleCloseModal}
+        >
+          Cancel
+        </Button>
```
We should also cancel the upload through websocket process on cancel. Currently after "cancel", the upload continues in the background and I think the success callback gets triggered.

mas-who replied:
@edlerd this is done, works nicely I think, thanks for the suggestion for removing the operation id, didn't even know that function existed

mas-who commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,349 @@
+  const handleCancelUpload = () => {
+    if (socket && socket.readyState === WebSocket.OPEN) {
+      socket.close(1000, CANCEL_UPLOAD_REASON);
+      eventQueue.remove(operationId);
```
yes good catch :+1: no need to check if operationId is defined though since Set.delete will work anyway in eventQueue

edlerd reviewed (commented):
QA LGTM
Code also looks good, added a couple of very tiny suggestions to improve on it slightly, should be good to go.

mas-who commented on src/pages/instances/forms/UploadExternalFormatFileForm.tsx:
```diff
@@ -0,0 +1,349 @@
+
+  const handleSubmit = (values: UploadExternalFormatFileFormValues) => {
+    if (!values.imageFile) {
+      return;
```
this is a unlikely condition since form submission is disabled if there is no file selected. I think I previously added this here to get around type checking, probably with some logic before the api call. I see that is no longer needed though, so I'd rather remove this condition than adding a notification.

mas-who commented on src/pages/instances/forms/UploadInstanceBackupFileForm.tsx:
```diff
@@ -0,0 +1,250 @@
+    const uploadController = new AbortController();
+    setUploadAbort(uploadController);
+
+    const instanceName = formik.values.name;
```
yeah not necessary

edlerd commented on src/pages/instances/forms/UploadInstanceBackupFileForm.tsx:
```diff
@@ -0,0 +1,250 @@
+
+  const handleUpload = () => {
+    if (!formik.values.instanceFile) {
+      return;
```
Perfect.

mas-who commented on src/util/uploadExternalFormatFile.tsx:
```diff
@@ -0,0 +1,165 @@
+  );
+};
+
+export const CANCEL_UPLOAD_REASON = "cancel upload";
```
no need :+1: I was trying to use this variable for something else previously and forgot to update it

mas-who commented on src/util/uploadExternalFormatFile.tsx:
```diff
@@ -0,0 +1,165 @@
+};
+
+// This method facilitates the uploading of a file to a server via a websocket connection.
+// 1. When the websocket connection opens, a FileReader instance is created to read the file in chunks.
+// 2. The readNextFileChunk function slices the file into chunks of the specified chunk size and converts them into binary buffers.
+// 3. If the chunk size is zero, indicating the end of the file, the websocket connection is closed.
+// 4. The FileReader.onload event handler processes indiviual file chunks as they get read, which sends them
+// through the websocket connection.
```
sure, it's also used to process data in chunks so we don't have to hold the entire file content in memory

edlerd reviewed (approved):
LGTM, good work :)
